src/session/session_dhandle.c

```diff
@@ -379,15 +378,16 @@ __session_dhandle_sweep(WT_SESSION_IMPL *session)
     TAILQ_FOREACH_SAFE(dhandle_cache, &session->dhandles, q, dhandle_cache_tmp)
     {
         dhandle = dhandle_cache->dhandle;
-        empty_btree = false;
-        if (WT_DHANDLE_BTREE(dhandle))
-            WT_WITH_DHANDLE(
-              session, dhandle, empty_btree = (__wt_btree_bytes_evictable(session) == 0));
 
+        /*
+         * Only discard handles that are dead or dying and, in the case of btrees, have been
+         * evicted. These checks are not done with any locks in place, other than the data handle
+         * reference, so we cannot peer past what is in the dhandle directly.
+         */
         if (dhandle != session->dhandle && dhandle->session_inuse == 0 &&
           (WT_DHANDLE_INACTIVE(dhandle) ||
-            (dhandle->timeofdeath != 0 && now - dhandle->timeofdeath > conn->sweep_idle_time) ||
-            empty_btree)) {
+            (dhandle->timeofdeath != 0 && now - dhandle->timeofdeath > conn->sweep_idle_time)) &&
+          (!WT_DHANDLE_BTREE(dhandle) || F_ISSET(dhandle, WT_DHANDLE_EVICTED))) {
```

Comment:
Earlier we did a session sweep if:
`in_use ==0 AND (handle_inactive OR handle_expired OR empty_btree)`

Now with the change, it looks like:
`in_use ==0 AND (handle_inactive OR handle_expired) AND (not-btree OR handle-marked-evicted)`

Effectively we are saying that for btree handles, the handle needs to be inactive or expired, but also it needs to have been marked evicted.

I don't see an issue with that, it seems correct to me. But not getting it right will prevent handles from being swept.

Let's do a second review too, just to make sure we have got this right.

Reply:
I think a review of the original ticket that added `empty_btree` might help drive whether it should be `and` or `or`.
Thanks for the good comment, as it hopefully makes clear so no one looks in the btree at a later date.